### ISSUE TITLE
Can't match @KeyFor to type

### ISSUE THREAD

**Issue by entonio** [typetools/checker-framework#1653]
Sorry for the title. I'm having trouble understanding this:

```java
public class Test {

    private Set<String> lastFound;

    private synchronized void abc(Object d) {

        Map<String, String> found = new HashMap<>();

        lastFound = found.keySet();
    }
}
```

results in
```
Test.java:9: warning: [assignment.type.incompatible] incompatible types in assignment.
        lastFound = found.keySet();
                                ^
  found   : Set<@KeyFor("found") String>
  required: Set<String>
```

What could the issue be?
Thank you.

**Comment by wmdietl** [comment]
This is caused by the Map Key Checker https://checkerframework.org/manual/#map-key-checker
`Map.keySet` returns a set of keys for that map, expressed as `@KeyFor("found")`.
Because generics are invariant in Java, the type `Set<@KeyFor("found") String>` can't be assigned to `Set<String>`.

You could add a cast, as in `(Set<String>) found.keySet();` ignoring that the elements are keys.
Or you change the field type to `Set<? extends String>`, which might be surprising, given that String is final. However, this allows the `@KeyFor` subtype for the elements.

Does the error make sense now?

**Comment by entonio** [comment]
To be honest, I've read it through and still don't get it. 
I understand why CF uses KeyFor, what I don't get is that it should complain about what in my view is correct code, nor can I see how the code could be improved in order to avoid the complaint.

It's also my policy not to introduce artefacts just to please CF (or any other tool), the most I do is

- add @Nullable/@NonNull casts (since those are CF-related anyway)
- or SuppressWarnings (and I do too many of those for may taste) with an explanation for what I think are false positives
- or sometimes reorganise the code slightly, but that doesn't include making changes that look strange or redundant

If this is something that can't be fixed, I'm ok with that, I'm just pointing out that it seems to me like an incorrect complaint.

**Comment by mernst** [comment]
The Checker Framework uses invariant subtyping for type parameters, just as Java does.

Do you understand why the following code:

``java
import java.util.Date;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

public class TestTypeParameterSubtyping {

    private Set<Date> lastFound;

    private synchronized void abc(Object d) {
        Map<Object, String> found = new HashMap<>();
        lastFound = found.keySet();
    }
}
```

results in this compilation error?

```
% javac -g TestTypeParameterSubtyping.java
TestTypeParameterSubtyping.java:12: error: incompatible types: Set<Object> cannot be converted to Set<Date>
        lastFound = found.keySet();
                                ^
1 error
```

(I'm asking this question to set the stage for answering your question.)

**Comment by entonio** [comment]
Assigning a Set<Object> to a Set<Date> could result in a cast error when extracting elements.

**Comment by mernst** [comment]
Right.  (Thanks!)

The cast error is just a symptom of the underlying problem, which is that there are two references to the same object with different types.  The two different references could be used in different ways -- one could insert values that are not legal for the other one.

The issue with your original code is the same.  Since `lastFound` has type `Set<String>` it is legal for code to put an arbitrary value in `lastFound`.  Doing so could corrupt the original object. 

That is correct behavior of the type-checker in general.  This explains why the Nullness Checker issues these warnings.

In this particular case, the code has specific peculiarities that make the code safe, such as:
 * Removal from `map.keySet()` changes `map`, and addition to `map` changes `map.keySet()`.
 * There are no uses of `lastFound` (no overriding, etc.).

(When understanding a Checker Framework error message, It's helpful to write an explanation of why the code is correct, and then see which step of your explanation the Checker Framework did not deduce on its own.  The Checker Framework manual has some advice about this:
https://checkerframework.org/manual/#handling-warnings .  Doing so also helps the maintainers know what you do and don't understand, and how best to respond to your query without wasting anyone's time.)

The Nullness Checker could be augmented with additional rules so that it did not issue a warning in this case.  I agree that this would make the checker more useful.

The rules would have to be constructed with care to avoid false negatives.  Maybe special handling of `keySet()` would be enough, but I haven't thought about it enough to be sure.  Whether it is worthwhile to make the enhancement depends in part on how often this case comes up.

Thanks for the report.

**Comment by entonio** [comment]
Hi,

> The cast error is just a symptom of the underlying problem, which is that there are two references to the same object with different types.
> 
Yes, I was simply pouring out why that would be a problem in your specific example. 

> The issue with your original code is the same. Since lastFound has type Set<String> it is legal for code to put an arbitrary value in lastFound.
> 
I don't follow. You're saying that since it is parametrised with a specific type, it's legal to put arbitrary values in it? Rather, it should only be legal to put `String` in it, which is the type parameter.

Your explanation would make sense to me if the types I've used for the Set and for the Map key weren't exactly the same (and I'm aware of the issues with subtypes and the like). But in this case they are exactly the same. You may have a point, but I'm just not getting the explanation, could you please add some code to my example that would showcase the issue?

Thank you.

**Comment by mernst** [comment]
> > The issue with your original code is the same. Since lastFound has type `Set<String>` it is legal for code to put an arbitrary value in lastFound.
>
> I don't follow. You're saying that since it is parametrised with a specific type, it's legal to put arbitrary values in it? Rather, it should only be legal to put `String` in it, which is the type parameter.

It is legal to put arbirtary `String` values in it.  It particular, it is legal to put non-`@KeyFor` values in it.

> Your explanation would make sense to me if the types I've used for the Set and for the Map key weren't exactly the same 

The type parameter in the return type of `m.keySet()` is [not the same](https://github.com/typetools/checker-framework/blob/master/checker/jdk/nullness/src/java/util/Map.java#L34) as the type paremeter in `m`'s type.  Maybe this is a cause of surprise or confusion.  It permits loops over `keySet()` to type-check.

Here is a concrete code example.  We don't want the code to run, because it would crash.
The two methods have the same code but different types, and it shows how the Nullness Checker doesn't permit the buggy code to type-check, regardless of the declared type of lst.


```java
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import org.checkerframework.checker.nullness.qual.KeyFor;

public class KeyTest {

  void method1(Map<String, Integer> m) {

    List<@KeyFor("m") String> lst = new ArrayList<>(m.keySet());

    @KeyFor("m") String key;

    key = lst.get(0);
    m.get(key).intValue();

    lst.add("some random string"); // Nullness Checker issues a warning

    key = lst.get(0);
    m.get(key).intValue();
  }

  void method2(Map<String, Integer> m) {

    List<String> lst = new ArrayList<>(m.keySet());

    @KeyFor("m") String key;

    key = lst.get(0);  // Nullness Checker issues a warning
    m.get(key).intValue();

    lst.add("some random string");

    key = lst.get(0); // Nullness Checker issues a warning
    m.get(key).intValue();
  }
}
```

**Comment by entonio** [comment]
I see. What I can say is:

- This seems to me like an abuse of the word 'type'. You're using it to restrict the instances, not the type. If you'd not conflate the two things from the start, then no explanation would have been needed. They may be types in CF-parlance, but they're not types in Java's mindset. I suppose you'll say an annotation automatically results in a subtype, but that's neither how programmers see their code, nor the real semantics you use seem to want it for.

- I don't see the usefulness of it, since you can't control the contents of the map. Any value you pick from its keySet may be removed at any time, resulting in a get returning null anyway. In your example there's only a crash because of the forced unboxing, I understand your ambition is to allow nonnull gets, but it looks way too flimsy for wide usage. I suppose that's what you mean when you say, in the documentation, that the Map checks aren't for every case, but their use by the nullness checker pretty much negates that.

- Whereas valid Java code with no artefacts can pass the nullness checker without issues, I srill don't see how to do it in this case. How does your proposal to change the field's type to `Set<? extends String>` make the code less buggy? In particular, how does such a solution sound like in light of the issues your use of @KeyFor attempts to solve?

I'm far from an expert in Java's types, much less generics. I fully believe you to be completely justified in your view of this issue, and myself unable to grasp it. However, I also think there ought to be a natural way to write the code of my example above that passes the nullness checker without issues.

**Comment by entonio** [comment]
(Note: `keySet` is not expected (by the average programmer) to return the map's own 'set of keys', but a copy of it. I'd even be surprised if it were mutable.)

**Comment by mernst** [comment]
We should clear up some misunderstandings.

> This seems to me like an abuse of the word 'type'. You're using it to restrict the instances, not the type.

That is excatly what a type is.  A standard definition of type is:  a set of values.  For example, the `int` type contains -2, -1, 0, 1, 2, 3, ... .  The `@Positive int` type contains `0, 1, 2, 3, ...`.  Type B is a subtype of type A exactly if type B is a subset of type A.  Writing a type on a variable is a promise that at run time the variable will have one of the values in that type.

> (Note: keySet is not expected (by the average programmer) to return the map's own 'set of keys', but a copy of it. I'd even be surprised if it were mutable.)

You are incorrect.  Every Java programmer expects `keySet()` to return a mutable view (not a copy) of the keys contained in the map, because Java programmers [read the Javadoc](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--).

> I don't see the usefulness of it, since you can't control the contents of the map. Any value you pick from its keySet may be removed at any time, resulting in a get returning null anyway. 

This is incorrect, per [the Javadoc for `keySet()`](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--).

**Comment by astiob** [comment]
Sorry for reviving an old ticket, but I just saw this due to its mention in the changelog. While I agree with most that you say, @mernst, I think your very last point may be a misunderstanding of what @entonio meant:

> I don't see the usefulness of it, since you can't control the contents of the map. Any value you pick from its keySet may be removed at any time, resulting in a get returning null anyway.

You can pick a key from a keySet and store it in a `@KeyFor` variable, then remove it from the map, and then try to use the previously assigned variable. Consider:

```java
class Test {
    void test(Map<String, String> map) {
        for (String key : map.keySet()) {
            map.remove(key);
            map.get(key).length();  // throws NPE
            break;
        }
    }
}
```

This code is clearly unsound, but the Checker Framework allows it to compile without so much as a warning (tested with 2.3.0). (`key` is implicitly `@KeyFor("map")` here, but the implicitness is not the problem: it could equally well be explicitly declared as such.)

Compare to the case when `key` is `@UnknownKeyFor`:

```java
class Test {
    String key = "";
    void test(Map<String, String> map) {
        for (String k : map.keySet()) {
            key = k;
            map.remove(key);
            map.get(key).length();
            break;
        }
    }
}
```

```
error: [dereference.of.nullable] dereference of possibly-null reference map.get(key)
            map.get(key).length();
                   ^
```

This does indeed seem to be a bit of a flaw in the map key checker… But I don’t immediately see how it could be fixed, short of making every `map.remove()` call invalidate all `@KeyFor("map")` qualifiers in the visible scope, reverting flow-refined locals to less informative types and producing a compilation error if a variable in scope has that qualifier explicitly specified, which seems too drastic. (And what about external code? I feel it might be possible to store an external reference to the same map and key qualified with `@KeyFor` that this wouldn’t invalidate.)

That said, the map key checker is certainly useful in other situations, so it seems there’s a certain soundness—usability trade-off here.

---

@entonio In case you read this:

> How does your proposal to change the field's type to `Set<? extends String>` make the code less buggy?

Your code in the original report declares the field as `Set<@UnknownKeyFor String>`, as `@UnknownKeyFor` is the default qualifier that denotes strings that may or may not be keys for any sets or maps. This means you are allowed to add arbitrary strings to the set. `@KeyFor("found") String` is a subtype of `@UnknownKeyFor String`, which should hopefully be easy to understand: after all, a string key for `found` is a particular kind of string (similarly to how `Integer` is a subtype of `Number`). If you were to save `found.keySet()` in the field and then try to add a string to it, you would be adding a value to the set that is not expected to be there, just as if you tried to add an arbitrary `Number` to a `Set<Integer>`.

The compiler (augmented with the Checker Framework’s map key checker) aims to detect this and reject such code. If you declare the field as `Set<? extends String>` (or `Set<? extends Number>` in the number analogy), then the rejection will come when you actually attempt to add the string to the set (this is standard Java compiler behaviour): the set could be a set of anything, and when you take out a value, you can be sure that it’s a `String` (or a `Number`), but you can’t add a new value because you don’t know what will fit inside and what won’t.

But if you declare the field as `Set<String>` (or `Set<Number>`), then the addition succeeds because you specifically state that the set does contain arbitrary strings (or numbers). However, if someone else creates a set of very particular strings (or very particular numbers) and passes it to you, they don’t want you to go adding arbitrary strings (or numbers) to it. Since storing this set in your field would allow you to do so, the compiler prevents you from doing it.

Now, in the particular case of `keySet()`, adding elements to the set is impossible anyway, but since Java doesn’t have a read-only set interface, this cannot be expressed more directly in Java. Essentially, declaring the set as `Set<? extends Something>` _is_ the Java way of expressing that the set can only be read from (and removed from) but not added to.
